File: frame/src/main/java/com/king/lib/frame/mvvm/MvvmActivity.java
package com.king.lib.frame.mvvm;

import android.os.Bundle;

import com.king.lib.frame.BaseCompactActivity;

import androidx.databinding.DataBindingUtil;
import androidx.databinding.ViewDataBinding;
import androidx.lifecycle.ViewModelProviders;

/**
 * Desc:
 *
 * @author：Jing Yang
 * @date: 2020/7/3 10:27
 */
public abstract class MvvmActivity<V extends ViewDataBinding, VM extends BaseViewModel> extends BaseCompactActivity {

    protected V mBinding;

    protected VM mModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mBinding = DataBindingUtil.setContentView(this, getContentView());
        mModel = createViewModel();
        if (mModel != null) {
            mModel.loadingObserver.observe(this, show -> {
                showLoading(show);
            });
//            mModel.messageObserver.observe(this, message -> showMessageLong(message));
        }
        onCreatePage(savedInstanceState);
    }

    protected void showLoading(Boolean show) {
        if (show) {
            showProgress();
        }
        else {
            dismissProgress();
        }
    }

    protected abstract int getContentView();

    protected abstract VM createViewModel();

    protected VM generateViewModel(Class<VM> vm) {
        return ViewModelProviders.of(this).get(vm);
    }

    protected abstract void onCreatePage(Bundle savedInstanceState);

}

File: frame_kotlin/src/main/java/com/king/lib/frame_kotlin/mvvm/MvvmActivity.kt
package com.king.lib.frame_kotlin.mvvm

import android.os.Bundle
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import com.king.lib.frame_kotlin.BaseCompactActivity
import io.reactivex.rxjava3.disposables.CompositeDisposable
import io.reactivex.rxjava3.disposables.Disposable

/**
 * 描述:
 *
 * 作者：景阳
 *
 * 创建时间: 2020/1/21 15:50
 */
abstract class MvvmActivity<T : ViewDataBinding, VM : BaseViewModel> : BaseCompactActivity() {

    lateinit var mBinding: T

    lateinit var mModel: VM

    var compositeDisposable: CompositeDisposable = CompositeDisposable()

    fun addDisposable(disposable: Disposable) {
        compositeDisposable.add(disposable)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mBinding = DataBindingUtil.setContentView(this, getContentView())
        mModel = createViewModel()
        if (mModel != null) {
            mModel!!.loadingObserver.observe(this, Observer { show ->
                if (show) {
                    showProgress()
                } else {
                    dismissProgress()
                }
            })
//            mModel!!.messageObserver.observe(this, Observer { message -> showMessageShort(message) })
        }

        onCreatePage(savedInstanceState)
    }

    abstract fun getContentView(): Int

    protected abstract fun createViewModel(): VM

    fun generateViewModel(vm: Class<VM>) = ViewModelProviders.of(this).get(vm)

    abstract fun onCreatePage(savedInstanceState: Bundle?)

    override fun onDestroy() {
        mModel?.onDestroy()
        super.onDestroy()
    }
}
